### PR TITLE
add mapcols! and repeat!, fix corner cases of repeat

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2192

Note that `repeat` in Base has a bug if `inner<0` or `outer<0` so I catch it now properly before we dispatch there

### CODE REVIEW

File: src/dataframe/dataframe.jl [hunk 1]
```diff
@@ -1368,3 +1368,79 @@ function Base.push!(df::DataFrame, row::Any)
     end
     df
 end
+
+"""
+    repeat!(df::DataFrame; inner::Integer = 1, outer::Integer = 1)
+
+Update a data frame `df` in place by repeating rows. `inner` specifies how many
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:1375]:
```suggestion
Update a data frame `df` in place by repeating its rows. `inner` specifies how many
```

File: src/dataframe/dataframe.jl [hunk 2]
```diff
@@ -1368,3 +1368,79 @@ function Base.push!(df::DataFrame, row::Any)
...
+function repeat!(df::DataFrame; inner::Integer = 1, outer::Integer = 1)
+    inner < 0 && throw(ArgumentError("inner keyword argument must be non-negative"))
+    outer < 0 && throw(ArgumentError("outer keyword argument must be non-negative"))
+    return mapcols!(x -> repeat(x, inner = inner, outer = outer), df)
+end
+
+"""
+    repeat!(df::DataFrame, count::Integer)
+
+Update a data frame `df` in place by repeating rows the number of times
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:1419]:
```suggestion
Update a data frame `df` in place by repeating its rows the number of times
```

File: test/utils.jl [hunk 3]
```diff
@@ -1,33 +1,84 @@
...
+
+    @test_throws MethodError repeat!(view(df, 1:2, :), 2)
+end
+
+@testset "repeat! inner_outer" begin
+    df = DataFrame(a = 1:2, b = 3:4)
+    ref = DataFrame(a = repeat(1:2, inner = 2, outer = 3),
+                    b = repeat(3:4, inner = 2, outer = 3))
+    repeat!(df, inner = 2, outer = 3)
+    @test df == ref
```
Comment (nalimilan) [inline test/utils.jl:64]: Also test that column vectors are `===` or `!==` from original column vectors (depending on our decision)?

**Comment by bkamins** [comment]
> OTOH it doesn't make much sense for performance.

Yes - that is why I have made it make a full copy. It is safer this way in case we have column aliases (we should not have typically, but in some cases we might have). And doing alias detection in this function would be an over-complication.

File: test/utils.jl (hunk 3 above)
Comment (bkamins) [reply test/utils.jl:70]: OK - added.

**Comment by bkamins** [comment]
@nalimilan this is not crucial to merge now, but IMHO is finished and rather simple to decide.

**Comment by bkamins** [comment]
I have added a note that `repeat!` freshly allocates columns.

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**